snuba/state/rate_limit.py

```diff
@@ -177,12 +177,25 @@ def rate_limit(
             )
         )
 
+    rate_limited = False
     try:
         yield stats
+        _, err, _ = sys.exc_info()
+        if isinstance(err, RateLimitExceeded):
```

Comment:
I am not a big fan of this approach. It seems that the exception is not propagated here because of the design of the ratelimitaggregator, thus we are breaking the abstraction to get the exception. 
I think this issue is pointing out that we either cannot model the rate limiter as a sequence of nested context managers, or that we should rebuild this as a context manager class where the `__exit__` method has information about the exception we exited with. 

I think the first may be a better option (avoid the nested rate limiters). 
What if this function applied a sequence of rate limiter settings instead of only one. 
It could delegate to another class (or two methods) the management of the state: add_query, push_query_back, remove_query, etc.
The responsibiltiy of this class would be to orchestrate the process:
- try to apply each rate limit setting. 
- if at least one fails, remove all
- if none fails yield 

Since this is not a fire right now. Why not actually making the design change ?

Reply:
Sure, I can rewrite it. I debated doing that restructuring to begin with but it seems like context managers are supposed to be the correct solution to this. The fact that `__exit__` is not called if an exception occurs in `__enter__` is bizarre to me and really makes this more complicated than it needs to be.

Reply:
After thinking about this, I actually think we should fix this as is for now. This will all be rewritten as part of the redesign effort and it seems like a waste of time to rewrite all this before then.

Reply:
Please add a TODO explaining that this will be refactored when rebuilding the rate limiter.